PhysicsTools/TensorFlowAOT/interface/Model.h

```diff
@@ -0,0 +1,161 @@
+#ifndef PHYSICSTOOLS_TENSORFLOWAOT_MODEL_H
+#define PHYSICSTOOLS_TENSORFLOWAOT_MODEL_H
+
+/*
+ * AOT model interface.
+ *
+ * Author: Marcel Rieger, Bogdan Wiederspan
+ */
+
+#include "PhysicsTools/TensorFlowAOT/interface/Util.h"
+#include "PhysicsTools/TensorFlowAOT/interface/Batching.h"
+
+namespace tfaot {
+
+  // model interface receiving the AOT wrapper type as a template argument
+  template <class W>
+  class Model {
+  public:
+    // constructor
+    explicit Model() : wrapper_(std::make_unique<W>()) {}
+
+    // destructor
+    ~Model(){};
+
+    // getter for the name
+    inline const std::string& name() const { return wrapper_->name(); }
+
+    // setter for the batch strategy
+    inline void setBatchStrategy(const BatchStrategy& strategy) { batchStrategy_ = strategy; }
+
+    // getter for the batch strategy
+    inline const BatchStrategy& getBatchStrategy() const { return batchStrategy_; }
+
+    // adds a new batch rule to the strategy
+    inline void setBatchRule(size_t batchSize, const std::vector<size_t>& sizes, size_t lastPadding = 0) {
+      batchStrategy_.setRule(BatchRule(batchSize, sizes, lastPadding));
+    }
+
+    // evaluates the model for multiple inputs and outputs of different types
+    template <typename... Outputs, typename... Inputs>
+    std::tuple<Outputs...> run(size_t batchSize, Inputs&&... inputs);
+
+  private:
+    std::unique_ptr<W> wrapper_;
```

Comment:
Since the wrapper contains the various models for different match sizes as members, and these models pre-allocate buffers for inputs, outputs and intermediate results, our intention was to avoid unwanted copying of the wrapper. But I guess with proper (move) assignment operators that you suggested in other comments, we can achieve the same.

Reply:
Ok. The use of `unique_ptr` here already prevents copying of `Model`, but that could be achieved also by deleting the copy constructor and assignment operator of `Model`.

Is `Model` expected to be moved around? If yes, keeping `unique_ptr` here would make sense if `W` would be "expensive to move". I see `Wrapper` base class has `std::string` (cheap to move) and `AllocMode` (presumably an `enum`, cheap to copy).

I see `wrapper.h.in` template to contain `unique_ptr` to the model class generated by TF (or something). Do I understand correctly that the class generated from `wrapper.h.in` would have one `unique_ptr` member for each batch size?

We probably don't want to be too dependent on how exactly the TF generated model class organizes it's data (whether it uses `std::vector` or `std::array`, or equivalent, data members), so probably one level of `unique_ptr` would indeed make sense to keep the moves cheap. Whether it should be here or in the `wrapper.h.in`-generated class is not immediately clear to me.

Reply:
We followed your recommendation in e08b8ba and kept the `unique_ptr` to the `Wrapper` in the `Model`, but the `Wrapper` contains the xla compiled function objects (we previously called this model as well, which we now changed to make this clearer) as normal members, resulting in just one level of pointers.